erxupdate/ntopnghosts.sh

```diff
@@ -0,0 +1,429 @@
+
+#functions
+date2stamp () { # source: https://www.unix.com/tips-and-tutorials/31944-simple-date-time-calulation-bash.html
+    if [[ "$1" == "now" ]]; then
+	    date --utc +%s
+	else
+		date --utc --date "$1" +%s
+	fi
+}
+
+dateDiff (){ # source: https://www.unix.com/tips-and-tutorials/31944-simple-date-time-calulation-bash.html
+    case $1 in
+        -s)   sec=1;      shift;;
+        -m)   sec=60;     shift;;
+        -h)   sec=3600;   shift;;
+        -d)   sec=86400;  shift;;
+        *)    sec=86400;;
+    esac
+    dte1=$(date2stamp "$1")
+    dte2=$(date2stamp "$2")
+    diffSec=$((dte2-dte1))
+    if ((diffSec < 0)); then abs=-1; else abs=1; fi
+    echo $((diffSec/sec*abs))
+}
+
+sendmessage (){
+	RESULT=`curl -s "https://api.telegram.org/bot$TELEGRAM_API_KEY/sendMessage?chat_id=$TELEGRAM_CHAT_ID&text=$MESSAGE"`
+	#echo "RESULT=$RESULT"
+	OK_STR=`echo $RESULT |jq -r '."ok"'`
+	#echo "OK_STR=$OK_STR"
+	if [[ "$OK_STR" != "true" ]]; then
+		# Failed to update:  We're going to wait 30 seconds and then retry and then we'll give up.  This means that if we fail again that there 
+		# will be no notification about this host
+		sleep 31s
+		RESULT=`curl -s "https://api.telegram.org/bot$TELEGRAM_API_KEY/sendMessage?chat_id=$TELEGRAM_CHAT_ID&text=$MESSAGE"`
+		OK_STR=`echo $RESULT |jq -r '."ok"'`
+	fi
+	if [[ "$OK_STR" == "true" ]]; then
+		OK_STR="OK"
+	fi
+	echo "Sending telegram message about $TELEGRAMHOSTNAME result=$OK_STR"
+}
+
+#initialise variables - this section shouldn't be modified
+# HOSTLOG is used for temporarily storing the records of hosts we see in case we need to write them to the hostnameregister if TELEGRAMENABLE is set to YES
+HOSTLOG=
+CRLF="\n"
+HOSTNAME=
+MAC=
+
+#check we're not already running
+if [[ "`ps -ef |grep ntopnghosts.sh | wc -l`" -gt "4" ]]; then
+    # if our process is present already we don't want to run again so we'll stop (more than 4 processes with our name means we are already running)
+	exit
+fi
+
+# we need to check if the hostnamesubstitute file needs to be updated
+if [[ "$NAMESUBSTITUTEURL" != "" ]]; then
+	NAMESUBSTITUTESIZE=`curl -sI $NAMESUBSTITUTEURL --location --silent -H 'Accept-Encoding: gzip,deflate' |grep -i content-length |cut -d ' ' -f2`
+	NAMESUBSTITUTESIZE=(${NAMESUBSTITUTESIZE:0:-1})
+	#curl -sI $NAMESUBSTITUTEURL --location --silent --write-out 'size_download=%{size_download}\n'
+	if [[ -f "/tmp/hostnamesubstitute" ]]; then
+		FILESIZE=`wc -c /tmp/hostnamesubstitute |cut -d ' ' -f1`
+	else
+		FILESIZE=0
+		if [[ -f "/etc/nprobe/hostnamesubstitute" ]]; then
+			cp /etc/nprobe/hostnamesubstitute /tmp/hostnamesubstitute
+		fi
+	fi
+	#echo "Centrally managed Namesubstitute size: $NAMESUBSTITUTESIZE"
+	#echo "Local Namesubstitute size: $FILESIZE"
+	if [[ "$NAMESUBSTITUTESIZE" != "0" ]]; then	
+		if [[ "$FILESIZE" != "$NAMESUBSTITUTESIZE" ]]; then
+			echo "Downloading updated name substitute file"
+			curl -s $NAMESUBSTITUTEURL -o "/tmp/hostnamesubstitute_new"
```

Comment:
This can open out to a series of security issues. I would at least force curl to use TLS. An attacker can trick you into storing data under tmp. Then, this data can be used later to inject shell code as it is not checked against any special character.

Reply:
updated curl commands to use --ssl
I don't want to use --ssl-reqd because if someone has a non-https server that they need to use then they would be unable to use the script.  I have added a note in the instructions recommending that https servers be used for hosting the script and hostnamesubstitute file

Reply:
I'll create a folder under /run to replace my use of /tmp.  This will ensure it is still on a tmpfs (RAM not long term storage to prevent wear of long term storage) and it will hopefully meet your requirement of not using the /tmp folder to reduce risk of attackers tricking us to using injected data in the tmp folder.